cmd/ttn-lw-stack/commands/as_db.go

```diff
@@ -0,0 +1,141 @@
+// Copyright © 2021 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package commands
+
+import (
+	"time"
+
+	"github.com/spf13/cobra"
+	"go.thethings.network/lorawan-stack/v3/pkg/applicationserver"
+	"go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io/packages"
+	asioapredis "go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io/packages/redis"
+	"go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io/pubsub"
+	asiopsredis "go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io/pubsub/redis"
+	"go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io/web"
+	asiowebredis "go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io/web/redis"
+	asredis "go.thethings.network/lorawan-stack/v3/pkg/applicationserver/redis"
+	"go.thethings.network/lorawan-stack/v3/pkg/identityserver/store"
+	"go.thethings.network/lorawan-stack/v3/pkg/redis"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+func toApplicationSet(list []*ttnpb.ApplicationIdentifiers) (set map[*ttnpb.ApplicationIdentifiers]struct{}) {
+	set = make(map[*ttnpb.ApplicationIdentifiers]struct{})
+	for _, ids := range list {
+		set[ids] = struct{}{}
+	}
+	return set
+}
+
+func toDeviceSet(list []*ttnpb.EndDeviceIdentifiers) (set map[*ttnpb.EndDeviceIdentifiers]struct{}) {
+	set = make(map[*ttnpb.EndDeviceIdentifiers]struct{})
+	for _, ids := range list {
+		set[ids] = struct{}{}
+	}
+	return set
+}
+
+var (
+	asDBCommand = &cobra.Command{
+		Use:   "as-db",
+		Short: "Manage Application Server database",
+	}
+	asDBCleanupCommand = &cobra.Command{
+		Use:   "cleanup",
+		Short: "Clean stale Application Server application data",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if config.Redis.IsZero() {
+				panic("Only Redis is supported by this command")
+			}
+
+			logger.Info("Connecting to Identity Server database...")
+			db, err := store.Open(ctx, config.IS.DatabaseURI)
```

Comment:
This will work in deployments where the Identity Server is local, but not in the ones where it is remote. I think we will have to add this method as an RPC that uses cluster authentication.

The component stores are fine (you may assume that you can just open them via Redis).

Reply:
I would say then this whole `as_db` command should be part of an AS RPC that is invoked from somewhere else. If we need to use cluster auth, we will have to do this procedure where we have access to the component and the running cluster.

But then do we invoke this as part of `ttn-lw-cli`? Or do we enhance `as_db` with an ability to attach to the current running component / cluster ?

Reply:
Pinging @htdvisser @johanstokking for thoughts on this. Making it an AS RPC would cause the performance of that particular AS instance to plummet most likely, which is why I preferred to have this range process separate from the real components, but at the same time calling the IS from a `-db` command is also not ideal either.